kernel/src/driver.rs

```diff
@@ -149,6 +149,19 @@ impl CommandResult {
     }
 }
 
+use core::convert::TryFrom;
+impl From<ReturnCode> for CommandResult {
```

Comment:
I see the point in adding a helper method for the common case, and especially for the first transition from the old to the new system call ABI. However, this is not the only way to encode a `ReturnCode` (and later `ErrorCode`) as a `CommandResult`. I'd like to emphasize this fact, and I believe that the generic `From` trait -- although common in the Rust ecosystem -- is poorly suited because of its semantic meaning elsewhere in the Rust ecosystem.

If we add a helper, I'd appreciate a more explicitly method on `CommandResult` (even `from_return_code`) to emphasize the fact that this is one particular way and helper to convert a `ReturnCode` to a `CommandResult` and by no means the only one (especially when other branches of a command would return other variants, given that each command is supposed to have only one success and failure variant).

Reply:
One `SuccessWithValue` is removed, how else would a `ReturnCode` be converted to a `CommandResult`? Or do you just mean that there might be places in Tock like this:

```rust
fn command(...) -> CommandResult {
    match command_num {
        0 => {
            let retcode = capsule_helper_func();
            let mut some_u32 = 0;
            if retcode == ReturnCode::SUCCESS {
                some_u32 = other_helper_func();
            }
            match retcode {
                ReturnCode::SUCCESS => CommandResult::success_u32(some_u32),
                _ => CommandResult::failure(),
            }
        }
    }
}
```
where because you want to return additional information, you have to construct the `CommandResult` differently?

Reply:
Or is this a concern about `?` automatically coercing when `From` is implemented? I don't think that can be an issue here either, because `CommandResult` is not actually a `Result<>` type?

Reply:
bors r+